src/lib.rs

```diff
@@ -250,64 +250,42 @@ impl BootInformationInner {
 
 impl fmt::Debug for BootInformation {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        writeln!(f, "multiboot information")?;
-
-        writeln!(
-            f,
-            "S: {:#010X}, E: {:#010X}, L: {:#010X}",
-            self.start_address(),
-            self.end_address(),
-            self.total_size()
-        )?;
-
-        if let Some(boot_loader_name_tag) = self.boot_loader_name_tag() {
-            writeln!(f, "boot loader name: {}", boot_loader_name_tag.name())?;
-        }
-
-        if let Some(command_line_tag) = self.command_line_tag() {
-            writeln!(f, "command line: {}", command_line_tag.command_line())?;
-        }
-
-        if let Some(memory_map_tag) = self.memory_map_tag() {
-            writeln!(f, "memory areas:")?;
-            for area in memory_map_tag.memory_areas() {
-                writeln!(
-                    f,
-                    "    S: {:#010X}, E: {:#010X}, L: {:#010X}",
-                    area.start_address(),
-                    area.end_address(),
-                    area.size()
-                )?;
-            }
-        }
-
-        if let Some(elf_sections_tag) = self.elf_sections_tag() {
-            writeln!(f, "kernel sections:")?;
-            for s in elf_sections_tag.sections() {
-                writeln!(
-                    f,
-                    "    name: {:15}, S: {:#08X}, E: {:#08X}, L: {:#08X}, F: {:#04X}",
-                    s.name(),
-                    s.start_address(),
-                    s.start_address() + s.size(),
-                    s.size(),
-                    s.flags().bits()
-                )?;
-            }
-        }
-
-        writeln!(f, "module tags:")?;
-        for mt in self.module_tags() {
-            writeln!(
-                f,
-                "    name: {:15}, S: {:#010X}, E: {:#010X}",
-                mt.name(),
-                mt.start_address(),
-                mt.end_address()
-            )?;
-        }
-
-        Ok(())
+        f.debug_struct("Multiboot2 Boot Information")
+            .field("start_address", &(self.start_address() as *const u64))
+            .field("end_address", &(self.end_address() as *const u64))
+            .field("total_size", &(self.total_size() as *const u64))
+            .field(
+                "boot_loader_name_tag",
+                &self
+                    .boot_loader_name_tag()
+                    .map(|x| x.name())
+                    .unwrap_or("<unknown>"),
+            )
+            .field(
+                "command_line",
+                &self
+                    .command_line_tag()
+                    .map(|x| x.command_line())
+                    .unwrap_or(""),
+            )
+            .field("memory_areas", &self.memory_map_tag())
+            // so far, I didn't found a nice way to connect the iterator with ".field()" because
+            // the iterator isn't Debug
+            .field("module_tags (count)", &self.module_tags().count())
```

Comment:
I would prefer an implementation of Debug for the iterator if possible. In my experience, having module tags printed helps debugging. Also, you could try mapping over the iterator to Debug each element?

Reply:
I absolutely agree with you. The behaviour should be equal to `Vec::iter()` which also prints all elements. But with the current implementation of the ModuleIter I can't achieve this because of different compile time errors, like lifetime and ownership problems :/